Add initial implementation of the year scroller

tehapo opened the pull request:
Fixes #35

This pull request introduces the parallax year scroller with several style and code changes.

 * Year scroller is a parallax scroller alongside the month scroller.
 * Year scroller is always visible if the viewport is wide enough.
 * If the year scroller is hidden, there are two ways toggle it open/closed.
  * Tap on the toolbar with the year number.
  * Swipe the month calendar to the left/right.
 * The "active" selection area of the scrollers is now the center of the scroller.
 * Tapping on the year scroller will do an animated scrolling to the tap position.
 * This PR also affects the styles widely, see the CSS changes. Notice especially following.
  * Fade to white effect on the month scroller when it has been slide to left.
  * Small indicator arrow on the year scroller.
  * Rotating arrow head icon on the toolbar to indicate open/closed state of the year scroller.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-date-picker/49)
<!-- Reviewable:end -->

tomivirkki commented:
*[vaadin-date-picker-overlay.html, line 16 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KB8MGQCu9NLNhtGarSG:-KB8MGQCu9NLNhtGarSH:606333646) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/c39477d01b65f9381ca91842b8b75783325c67e9/vaadin-date-picker-overlay.html#L16)):*
Change to 100%

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 161 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KB8R2H2mXU56Y8uzcys:-KB8R2H2mXU56Y8uzcyt:-2007028427) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/c39477d01b65f9381ca91842b8b75783325c67e9/vaadin-date-picker-overlay.html#L161)):*
"hidden" requires explicit iron-flex-layout import & dependency....hmm...does it make sense to add it just for this one case?

---

*[vaadin-infinite-scroller.html, line 117 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KB8VGYiI1rZ6JWSSi7T:-KB8VGYjbRbqly9E9qfx:1232660049) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/c39477d01b65f9381ca91842b8b75783325c67e9/vaadin-infinite-scroller.html#L117)):*
I'd rather do this in the client rather than adding it to the public api of this component. In the end of the day it doesn't really offer anything a client couldn't do with the existing api (position). (Notice that you can't use scrollTop here because it doesn't work with big scrolls -> use position API instead.) If you really want to add this to the component's public api, then it requires tests and all.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49)*
<!-- Sent from Reviewable.io -->

tehapo commented:
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 16 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KB8MGQCu9NLNhtGarSG:-KBCJdthfz6zM42OwFPY:-499146009) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/c39477d01b65f9381ca91842b8b75783325c67e9/vaadin-date-picker-overlay.html#L16)):*
Done.

---

*[vaadin-date-picker-overlay.html, line 161 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KB8R2H2mXU56Y8uzcys:-KBCJg7FsnSoG_m7J39d:233239993) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/c39477d01b65f9381ca91842b8b75783325c67e9/vaadin-date-picker-overlay.html#L161)):*
Added a `display: none` for it in the element styles. So this will now work even without iron-flex-layout present.

---

*[vaadin-infinite-scroller.html, line 117 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KB8VGYiI1rZ6JWSSi7T:-KBCJoMuNmKFrLeYfpbb:-1623545231) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/c39477d01b65f9381ca91842b8b75783325c67e9/vaadin-infinite-scroller.html#L117)):*
You're right, I agree. Moved the logic into `vaadin-date-picker-overlay` and added a quick easing function there to make it more smooth. This will most likely need refactoring to be used with the today button, but decided to keep that change separate.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49)*
<!-- Sent from Reviewable.io -->

platosha commented:
Reviewed 6 of 8 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 26 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KBCnpun5VvX-rm7LYzR:-KBCnpun5VvX-rm7LYzS:-1496984426) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/091c826ba6770a613b66ed1a2a99f4000900f936/vaadin-date-picker-overlay.html#L26)):*
Not a blocker, but personally I don't like to have relative font size items. It drops from typographic font sizes scale defined in material design guidelines, and results in a fractional font-size, which is also not nice at all (it makes all `em` items to be in fractional pixels).

You're already included `"paper-styles/typography.html"`, why not pick matching sizes from these mixins then?

If don't like to use the mixins, would be still better to have a pixel-accurate font size from the material design scale here. Either with fixed px-value, or at least a relative size that would by default result in a pixel-accurate value.

See: https://www.google.com/design/spec/style/typography.html#typography-styles

---

*[vaadin-date-picker-overlay.html, line 221 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KBCqvwX-BA1LdeQXIlv:-KBCqvwX-BA1LdeQXIlw:-1976184394) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/091c826ba6770a613b66ed1a2a99f4000900f936/vaadin-date-picker-overlay.html#L221)):*
This would set the `_desktopMode` once on init, and would not change ever since then. Should we check the window width more often than just once during initialization? 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49)*
<!-- Sent from Reviewable.io -->

Saulis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 221 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KBCqvwX-BA1LdeQXIlv:-KBCoRwSaz6l8gk9oCZR:1640063455) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/091c826ba6770a613b66ed1a2a99f4000900f936/vaadin-date-picker-overlay.html#L221)):*
Would like to try using <iron-media-query> to set this property, but OK if we do it in another ticket later.

---

*[vaadin-month-calendar.html, line 128 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KBCpssDxwehceUBf0yT:-KBCpssEaPKFzigfias8:-301332979) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/091c826ba6770a613b66ed1a2a99f4000900f936/vaadin-month-calendar.html#L128)):*
This is :+1: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-:-KBDArQrEqzXRM_RGCzp:980077423)*
<!-- Sent from Reviewable.io -->

tehapo commented:
Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-date-picker-overlay.html, line 26 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KBCnpun5VvX-rm7LYzR:-KBDC37tmfHhY_1-txJI:1445463193) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/091c826ba6770a613b66ed1a2a99f4000900f936/vaadin-date-picker-overlay.html#L26)):*
Done. Switched to using font definitions from the `typography.html`.

---

*[vaadin-date-picker-overlay.html, line 221 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49#-KBCqvwX-BA1LdeQXIlv:-KBDC4O3By6GiWGoCKwa:-1458959882) ([raw file](https://github.com/vaadin/vaadin-date-picker/blob/091c826ba6770a613b66ed1a2a99f4000900f936/vaadin-date-picker-overlay.html#L221)):*
Let's postpone this to another ticket. Would be nice to merge this one ASAP to get the other PRs rebased.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49)*
<!-- Sent from Reviewable.io -->

platosha commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-date-picker/49)*
<!-- Sent from Reviewable.io -->
